File: apps/go/internal/chapter15/api_test.go
package chapter15_test

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter15"
)

// stubStore は置き場のスタブ。学習しなくても API の振る舞いを確かめられる。
type stubStore struct {
	sales    chapter15.SalesModel
	survival chapter15.SurvivalModel
}

func (s stubStore) LoadSalesModel() (chapter15.SalesModel, error) {
	if s.sales == nil {
		return nil, chapter15.ModelNotFound(chapter15.SalesModelName)
	}

	return s.sales, nil
}

func (s stubStore) LoadSurvivalModel() (chapter15.SurvivalModel, error) {
	if s.survival == nil {
		return nil, chapter15.ModelNotFound(chapter15.SurvivalModelName)
	}

	return s.survival, nil
}

// fixedSales は決まった値を返す興行収入のモデル。
type fixedSales struct {
	sales float64
}

func (m fixedSales) PredictSales(chapter15.Movie) (float64, error) { return m.sales, nil }

// fixedSurvival は決まった値を返す生存予測のモデル。
type fixedSurvival struct {
	survived bool
}

func (m fixedSurvival) Survives(chapter15.Passenger) (bool, error) { return m.survived, nil }

// call は API を呼んで、応答の状態コードと本文を返す。
func call(t *testing.T, handler http.Handler, method, path, body string) (int, string) {
	t.Helper()

	request := httptest.NewRequest(method, path, strings.NewReader(body))
	recorder := httptest.NewRecorder()
	handler.ServeHTTP(recorder, request)

	return recorder.Code, recorder.Body.String()
}

// handlerWith はスタブの置き場を差し込んだ API を作る。
func handlerWith(store chapter15.ModelStore) http.Handler {
	return chapter15.NewHandler(chapter15.NewPredictionService(store))
}

// bothModels は両方のモデルがあるスタブ。
func bothModels() stubStore {
	return stubStore{sales: fixedSales{sales: 4321.5}, survival: fixedSurvival{survived: true}}
}

func TestPredictionAPI(t *testing.T) {
	t.Parallel()

	t.Run("興行収入を予測して JSON で返す", func(t *testing.T) {
		t.Parallel()

		status, body := call(t, handlerWith(bothModels()), http.MethodPost, "/cinema/sales",
			`{"sns1": 100, "sns2": 2000, "actor": 300, "original": 1}`)

		if status != http.StatusOK {
			t.Fatalf("状態コード = %d, want %d（本文 %s）", status, http.StatusOK, body)
		}

		var response struct {
			Sales float64 `json:"sales"`
		}
		if err := json.Unmarshal([]byte(body), &response); err != nil {
			t.Fatalf("応答が JSON ではありません: %v", err)
		}

		if response.Sales != 4321.5 {
			t.Errorf("sales = %v, want %v", response.Sales, 4321.5)
		}
	})

	t.Run("生存を予測して JSON で返す", func(t *testing.T) {
		t.Parallel()

		status, body := call(t, handlerWith(bothModels()), http.MethodPost, "/survived",
			`{"pclass": 1, "sex": "female", "age": 30, "sib_sp": 0, "parch": 0, "fare": 80, "embarked": "S"}`)

		if status != http.StatusOK {
			t.Fatalf("状態コード = %d, want %d（本文 %s）", status, http.StatusOK, body)
		}

		if !strings.Contains(body, `"survived":true`) {
			t.Errorf("本文 = %s, want survived が true", body)
		}
	})

	t.Run("年齢と乗船港は省略できる", func(t *testing.T) {
		t.Parallel()

		status, body := call(t, handlerWith(bothModels()), http.MethodPost, "/survived",
			`{"pclass": 3, "sex": "male", "sib_sp": 0, "parch": 0, "fare": 7.25}`)

		if status != http.StatusOK {
			t.Errorf("状態コード = %d, want %d（本文 %s）", status, http.StatusOK, body)
		}
	})

	t.Run("入力が不正なら 422 と理由の一覧を返す", func(t *testing.T) {
		t.Parallel()

		tests := []struct {
			name string
			path string
			body string
			want []string
		}{
			{
				name: "映画の必須の値が無い",
				path: "/cinema/sales",
				body: `{"sns1": 100}`,
				want: []string{"sns2 は必須です", "actor は必須です", "original は必須です"},
			},
			{
				name: "映画の値が負で、原作の有無が選択肢にない",
				path: "/cinema/sales",
				body: `{"sns1": -1, "sns2": 2000, "actor": 300, "original": 2}`,
				want: []string{"sns1 は 0 以上にしてください", "original は 0、1 のどれかにしてください"},
			},
			{
				name: "乗客の等級と性別が選択肢にない",
				path: "/survived",
				body: `{"pclass": 4, "sex": "unknown", "sib_sp": 0, "parch": 0, "fare": 10}`,
				want: []string{"pclass は 1、2、3 のどれかにしてください", "sex は female、male のどれかにしてください"},
			},
			{
				name: "乗船港が選択肢にない",
				path: "/survived",
				body: `{"pclass": 1, "sex": "male", "sib_sp": 0, "parch": 0, "fare": 10, "embarked": "X"}`,
				want: []string{"embarked は C、Q、S のどれかにしてください"},
			},
		}

		for _, test := range tests {
			t.Run(test.name, func(t *testing.T) {
				t.Parallel()

				status, body := call(t, handlerWith(bothModels()), http.MethodPost, test.path, test.body)

				if status != http.StatusUnprocessableEntity {
					t.Fatalf("状態コード = %d, want %d（本文 %s）", status, http.StatusUnprocessableEntity, body)
				}

				for _, want := range test.want {
					if !strings.Contains(body, want) {
						t.Errorf("本文 = %s, want %q を含む", body, want)
					}
				}
			})
		}
	})

	t.Run("JSON として読めなければ 422 を返し、内部の型名を漏らさない", func(t *testing.T) {
		t.Parallel()

		for _, body := range []string{`{"sns1": "たくさん"}`, `{ここは JSON ではない`, ``} {
			status, response := call(t, handlerWith(bothModels()), http.MethodPost, "/cinema/sales", body)

			if status != http.StatusUnprocessableEntity {
				t.Errorf("%q の状態コード = %d, want %d", body, status, http.StatusUnprocessableEntity)
			}

			if !strings.Contains(response, "JSON の形式または値の型が正しくありません") {
				t.Errorf("本文 = %s, want 形式の理由", response)
			}

			if strings.Contains(response, "chapter15") || strings.Contains(response, "float64") {
				t.Errorf("本文 = %s, want 内部の型名を含まない", response)
			}
		}
	})

	t.Run("モデルが無ければ 503 を返す", func(t *testing.T) {
		t.Parallel()

		status, body := call(t, handlerWith(stubStore{}), http.MethodPost, "/cinema/sales",
			`{"sns1": 100, "sns2": 2000, "actor": 300, "original": 1}`)

		if status != http.StatusServiceUnavailable {
			t.Fatalf("状態コード = %d, want %d（本文 %s）", status, http.StatusServiceUnavailable, body)
		}

		if !strings.Contains(body, "モデルがありません: cinema") {
			t.Errorf("本文 = %s, want モデルの名前を含む", body)
		}
	})

	t.Run("ヘルスチェックはモデルごとの状態を返す", func(t *testing.T) {
		t.Parallel()

		tests := []struct {
			name   string
			store  stubStore
			status string
		}{
			{name: "両方ある", store: bothModels(), status: "ok"},
			{name: "片方だけある", store: stubStore{sales: fixedSales{}}, status: "degraded"},
			{name: "どちらも無い", store: stubStore{}, status: "degraded"},
		}

		for _, test := range tests {
			t.Run(test.name, func(t *testing.T) {
				t.Parallel()

				status, body := call(t, handlerWith(test.store), http.MethodGet, "/health", "")

				if status != http.StatusOK {
					t.Fatalf("状態コード = %d, want %d", status, http.StatusOK)
				}

				var response struct {
					Status string          `json:"status"`
					Models map[string]bool `json:"models"`
				}
				if err := json.Unmarshal([]byte(body), &response); err != nil {
					t.Fatalf("応答が JSON ではありません: %v", err)
				}

				if response.Status != test.status {
					t.Errorf("status = %q, want %q", response.Status, test.status)
				}

				if len(response.Models) != 2 {
					t.Errorf("models = %v, want 2 件", response.Models)
				}
			})
		}
	})

	t.Run("知らないパスは 404、許していないメソッドは 405 を返す", func(t *testing.T) {
		t.Parallel()

		handler := handlerWith(bothModels())

		if status, _ := call(t, handler, http.MethodGet, "/unknown", ""); status != http.StatusNotFound {
			t.Errorf("状態コード = %d, want %d", status, http.StatusNotFound)
		}

		if status, _ := call(t, handler, http.MethodGet, "/cinema/sales", ""); status != http.StatusMethodNotAllowed {
			t.Errorf("状態コード = %d, want %d", status, http.StatusMethodNotAllowed)
		}
	})
}

File: apps/clojure/test/getting_started_ml/chapter15/api_test.clj
(ns getting-started-ml.chapter15.api-test
  "第 15 章: 予測 API のテスト。ハンドラーは関数なので、サーバーを起動せずに直接呼ぶ。"
  (:require [clojure.test :refer [deftest is testing]]
            [getting-started-ml.chapter15.api :as api]
            [getting-started-ml.chapter15.domain :as domain]
            [getting-started-ml.chapter15.fakes :as fakes]))

(defn- request
  "Ring の要求のマップを作る。本文は入力ストリームで渡す（Jetty が渡すものと同じ形）。"
  ([method uri] {:request-method method :uri uri})
  ([method uri body]
   (assoc (request method uri)
          :body (java.io.ByteArrayInputStream. (.getBytes ^String body "UTF-8")))))

(defn- call
  "偽物の置き場を使うハンドラーを、1 回の要求で呼ぶ。"
  [store-options request-map]
  ((api/handler (fakes/fake-store store-options)) request-map))

(def ^:private ready {:sales true :survival true})

(deftest ヘルスチェック
  (testing "すべて読み込めれば ok を返す"
    (let [response (call ready (request :get "/health"))]
      (is (= 200 (:status response)))
      (is (= "application/json; charset=utf-8" (get-in response [:headers "Content-Type"])))
      (is (= "{\"status\":\"ok\",\"models\":{\"cinema\":true,\"survived\":true}}" (:body response)))))
  (testing "読み込めないモデルがあれば degraded を返す"
    (is (= "{\"status\":\"degraded\",\"models\":{\"cinema\":true,\"survived\":false}}"
           (:body (call {:sales true :survival false} (request :get "/health")))))))

(deftest 興行収入を予測する
  (let [response (call ready (request :post "/cinema/sales"
                                      "{\"sns1\": 100, \"sns2\": 2000, \"actor\": 300, \"original\": 1}"))]
    (is (= 200 (:status response)))
    (is (= "{\"sales\":4321.5}" (:body response)))))

(deftest 生存を予測する
  (let [response (call ready (request :post "/survived"
                                      "{\"pclass\": 1, \"sex\": \"female\", \"sib_sp\": 0, \"parch\": 0, \"fare\": 80}"))]
    (is (= 200 (:status response)))
    (is (= "{\"survived\":true}" (:body response)))))

(deftest 不正な入力は四百二十二を返す
  (testing "検証の理由を並べて返す"
    (let [response (call ready (request :post "/cinema/sales"
                                        "{\"sns1\": -1, \"sns2\": 2000, \"actor\": 300, \"original\": 2}"))]
      (is (= 422 (:status response)))
      (is (= "{\"detail\":[\"sns1 は 0 以上にしてください\",\"original は 0、1 のどれかにしてください\"]}"
             (:body response)))))
  (testing "JSON として読めない入力も四百二十二にする"
    (let [response (call ready (request :post "/cinema/sales" "{\"sns1\": \"たくさん\"}"))]
      (is (= 422 (:status response)))
      (is (= "{\"detail\":[\"JSON の形式または値の型が正しくありません\"]}" (:body response))))))

(deftest モデルが無ければ五百三を返す
  (let [response (call {:sales false :survival false}
                       (request :post "/cinema/sales"
                                "{\"sns1\": 100, \"sns2\": 2000, \"actor\": 300, \"original\": 1}"))]
    (is (= 503 (:status response)))
    (is (= "{\"detail\":\"学習済みモデル cinema が見つかりません\"}" (:body response)))))

(deftest 予測が失敗すれば五百を返す
  ;; 置き場が約束と違う例外を投げる。応答には内部の事情を出さない
  (let [broken (reify domain/ModelStore
                 (load-sales-model [_] (throw (RuntimeException. "内部の秘密")))
                 (load-survival-model [_] (throw (RuntimeException. "内部の秘密"))))
        response ((api/handler broken)
                  (request :post "/cinema/sales"
                           "{\"sns1\": 100, \"sns2\": 2000, \"actor\": 300, \"original\": 1}"))]
    (is (= 500 (:status response)))
    (is (= "{\"detail\":\"予測できませんでした\"}" (:body response)))))

(deftest 知らないパスは四百四_許していないメソッドは四百五
  (let [not-found (call ready (request :get "/unknown"))
        not-allowed (call ready (request :get "/cinema/sales"))]
    (is (= 404 (:status not-found)))
    (is (= "{\"detail\":\"見つかりません\"}" (:body not-found)))
    (is (= 405 (:status not-allowed)))
    (is (= "POST" (get-in not-allowed [:headers "Allow"])))
    (is (= "{\"detail\":\"許していないメソッドです\"}" (:body not-allowed)))))
